### PR TITLE
Enhance Gameplay Reliability with Offline Mode, Automatic Reconnection, and Save File Synchronization

### PR DESCRIPTION
**Pull request by f-fsantos**
## What are the changes?
Previously, losing connection to the server interrupted gameplay. With these changes:
- Players can continue offline if the server is down.
- The game periodically checks server status and reconnects automatically.
- Reconnection is smooth, without needing a page reload.
- A banner shows the current server status.
- Players can refresh the page and still play offline if they connected at least once.
- Save files are synchronized with the server when it becomes available again.

## Why am I making these changes?
These improvements aim to provide an uninterrupted gaming experience, even during server downtimes or DDoS attacks. By enabling offline play, smooth reconnections, and save file synchronization, we ensure players can continue their game without disruptions, enhancing overall user satisfaction.

## How to test the changes?
To test:
1. Disconnect from the internet and start the game to verify offline functionality.
2. Reconnect to the internet and check for automatic server reconnection without page reload.
3. Verify the banner shows server status.
4. Refresh the page after connecting once and confirm offline functionality.
5. Ensure save files synchronize with the server when it comes back online.

## Checklist
- [x] Have I checked for no overlap with another PR?
- [x] Is the PR self-contained and not splittable into smaller PRs?
- [x] Have I clearly explained the changes?
- [x] Have I tested the changes manually?
- [x] Are there visual changes?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/system/game-data.ts [hunk 1]
```diff
@@ -623,9 +623,9 @@ export class GameData {
       pokeballCounts: scene.pokeballCounts,
       money: scene.money,
       score: scene.score,
-      waveIndex: scene.currentBattle.waveIndex,
-      battleType: scene.currentBattle.battleType,
-      trainer: scene.currentBattle.battleType === BattleType.TRAINER ? new TrainerData(scene.currentBattle.trainer) : null,
+      waveIndex: scene.currentBattle?.waveIndex,
```
Comment (Greenlamp2) [inline src/system/game-data.ts:632]: ? not needed anymore

File: src/system/game-data.ts [hunk 2]
```diff
@@ -623,9 +623,9 @@ export class GameData {
...
+      battleType: scene.currentBattle?.battleType,
```
Comment (Greenlamp2) [inline src/system/game-data.ts:633]: ? not needed anymore

File: src/system/game-data.ts [hunk 3]
```diff
@@ -623,9 +623,9 @@ export class GameData {
...
+      trainer: scene.currentBattle?.battleType === BattleType.TRAINER ? new TrainerData(scene.currentBattle?.trainer) : null,
```
Comment (Greenlamp2) [inline src/system/game-data.ts:634]: ? not needed anymore

File: index.css [hunk 4]
```diff
@@ -175,4 +175,15 @@ body {
 input:-internal-autofill-selected {
   -webkit-background-clip: text;
   background-clip: text;
+}
+#banner {
+  position: fixed;
```
Comment (Greenlamp2) [inline index.css:180]: we can put absolute there and remove it from the js code

**Comment by flx-sta** [comment]
Feel free to reopen it when you come back to it

**Pull request closed without merging**